node/network/connection-request-issuer/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This subsystem is responsible for keeping track of session changes
+//! and issuing a connection request to the validators relevant to
+//! the gossiping subsystems on every new session.
+
+use futures::FutureExt as _;
+use polkadot_node_subsystem::{
+	messages::{
+		ConnectionRequestIssuerMessage,
+	},
+	ActiveLeavesUpdate, FromOverseer, OverseerSignal,
+	Subsystem, SpawnedSubsystem, SubsystemContext,
+};
+use polkadot_node_subsystem_util::{
+	validator_discovery::{ConnectionRequest, self},
+	self as util,
+};
+use polkadot_primitives::v1::{
+	Hash, ValidatorId, SessionIndex,
+};
+use polkadot_node_network_protocol::peer_set::PeerSet;
+
+const LOG_TARGET: &str = "connection_request_issuer";
+
+/// The Connection Request Issuer subsystem.
+pub struct ConnectionRequestIssuer {}
+
+#[derive(Default)]
+struct State {
+	last_session_index: Option<SessionIndex>,
+	/// when we overwrite this, it automatically drops the previous request
+	last_connection_request: Option<ConnectionRequest>,
+}
+
+impl ConnectionRequestIssuer {
+	/// Create a new instance of the [`ConnectionRequestIssuer`] subsystem.
+	pub fn new() -> Self {
+		Self {}
+	}
+
+	#[tracing::instrument(skip(self, ctx), fields(subsystem = LOG_TARGET))]
+	async fn run<Context>(self, ctx: Context)
+	where
+		Context: SubsystemContext<Message = ConnectionRequestIssuerMessage>,
+	{
+		let mut state = State::default();
+		self.run_inner(ctx, &mut state).await
+	}
+
+	#[tracing::instrument(skip(self, ctx, state), fields(subsystem = LOG_TARGET))]
+	async fn run_inner<Context>(self, mut ctx: Context, state: &mut State)
+	where
+		Context: SubsystemContext<Message = ConnectionRequestIssuerMessage>,
+	{
+		loop {
+			let message = match ctx.recv().await {
+				Ok(message) => message,
+				Err(e) => {
+					tracing::debug!(
+						target: LOG_TARGET,
+						err = ?e,
+						"Failed to receive a message from Overseer, exiting"
+					);
+					return;
+				},
+			};
+			match message {
+				FromOverseer::Communication {
+					msg: _,
+				} => {}
+				FromOverseer::Signal(OverseerSignal::ActiveLeaves(ActiveLeavesUpdate {
+					activated,
+					..
+				})) => {
+					tracing::trace!(target: LOG_TARGET, "active leaves signal");
+
+					let leaves = activated.into_iter().map(|(h, _)| h);
+					if let Err(e) = state.handle_active_leaves(&mut ctx, leaves).await {
+						tracing::debug!(target: LOG_TARGET, "Error {}", e);
+					}
+				}
+				FromOverseer::Signal(OverseerSignal::BlockFinalized(_hash, _number)) => {},
+				FromOverseer::Signal(OverseerSignal::Conclude) => {
+					return;
+				}
+			}
+		}
+	}
+}
+
+async fn determine_relevant_validators(
+	ctx: &mut impl SubsystemContext,
+	relay_parent: Hash,
+	_session: SessionIndex,
+) -> Result<Vec<ValidatorId>, util::Error> {
+	let validators = util::request_validators_ctx(relay_parent, ctx).await?.await??;
+	Ok(validators)
```

Comment:
Yeah, when just using all validators, we should definitely not use the reserved set. As we don't need gossip, if we send to all peers directly anyway.

Reply:
Yes, I was planning to address it in a separate PR (see p.2 of the issue) and get something working first.